### ISSUE TITLE
Add command-line interface

### ISSUE THREAD

**Issue by jbingham** [ohsu-comp-bio/funnel#38]
One thing we've found super useful at Google/Verily is to have a qsub-style command-line interface to the Pipelines API. We've got a version called dsub now at:

https://github.com/googlegenomics/task-submission-tools

We'd like to enhance dsub so that it supports TES as a backend. Then it could support local execution, job schedulers on HPC clusters, and multiple cloud providers.

Some possibilities:
- we could add Funnel as a backend to dsub. 
- we could fork dsub and add it as a CLI to Funnel
- we could contribute dsub to the ga4gh repo once it supports pluggable backends

**Comment by buchanae** [comment]
A TES/funnel backend for dsub sounds great.

Ideally, I'd like to have a single funnel binary that includes something like dsub. That would go towards the goal of super easy download and usage.

A crazy idea (and probably separate issue): it would be amazing if somehow I could run `funnel run gcr.io/my-container samtools index gs://path/to/input.bam gs://path/to/output.bai`. I don't know how you'd determine which paths are outputs, but it might be possible.

**Comment by jbingham** [comment]
The current dsub command-line would be:

dsub --project bill-to-project --logging gs://path/to/logs --image gcr.io/my-container --command "samtools index ${BAM} ${BAI}" --input BAM=gs://path/to/input.bam --output BAI=gs://path/to/output.bai

It's not too far off. With a few tweaks, the billing project could be scooped up from the user's environment. The logging path could default to the output path gs://path/to/.

**Comment by buchanae** [comment]
I've been experimenting with a `funnel run` command here:
https://github.com/buchanae/funnel/tree/tes-client

The current syntax is something like:
`funnel run ubuntu 'md5sum {in .SRC} > {out .DST}' -- -SRC ~/path/to/input.txt -DST gs://bkt/out.txt`

This would send the task to server and print a job ID to stdout. There's a `--wait` flag for blocking until the job is done. There's a `--print` flag for printing the task JSON without running it. There are some flags for setting the resource requests (cpu, ram, disk, etc).

Being able to map/scatter over lines in a TSV/JSON/YAML file would be useful. That would print out multiple job IDs, one per line.

A `funnel wait` command might also be useful for waiting on a group of job IDs, which might give you more flexibility in throwing together ad hoc pipelines.

The code is still pretty rough and experimental and full of holes, but with a little work I think it can be pretty useful.

Call/task caching would make `funnel run` and `dsub` commands really useful, I think. I imagine having a bash script with a bunch of these commands, which I can run repeatedly without worrying about rerunning heavyweight processing tasks. It would be a sort of poor man's workflow engine, which sounds really useful. </yet another crazy idea>

**Comment by jbingham** [comment]
Your command-line is functionally equivalent to the dsub version:

dsub --image ubuntu --command 'md5sum ${IN} > ${OUT}' --input IN=gs://foo/input.txt --output OUT=gs://foo/out.txt

In dsub, we also have a --wait flag -- not a separate command, just a flag.

dsub --image ubuntu --command 'md5sum ${IN} > ${OUT}' --input IN=gs://foo/input.txt --output OUT=gs://foo/out.txt --wait

Now the command will block until completion.

**Comment by buchanae** [comment]
After trying to implement the details in the design mentioned in a previous comment,
I ended up moving closer to the dsub design. The commands below are a rough sketch,
and is not implemented yet.

```
funnel run  \
    'bowtie2 -f $factor -x $other -p1 $pair1 -p2 $pair2 -o $alignments' \
    --in pair1 ~/pair1.fastq                        \
    --in pair2 ~/pair2.fastq                        \
    --contents other ~/input.txt                    \
    --name pair1 'Pair 1 inputs'                    \
    --desc pair1 'Pair 1 description'               \
    --type pair1 directory                          \
    --out alignments gs://bkt/bowtie/alignments.bam \
    --factor 5                                      \
    --image my-bowtie2                              \
    --name 'Bowtie2 test'                           \
    --volume /tmp                                   \
    --volume /opt                                   \
    --tag 'foo: bar'                                \
    --description 'Testings an example of using `funnel run` for a bowtie2 command'

  funnel run <<TPL
    --cpus 10
    --preemptible
    --ram 150GB
    --disk 1TB
    --zones west,east
  TPL

  align=$(cat <<TPL
    'bowtie2 -f $factor -x $other -p1 $pair1 -p2 $pair2 -o $alignments' 
    -i pair1 ~/pair1.fastq            
    -i pair2 ~/pair2.fastq           
    -o alignments gs://bkt/bowtie/alignments.bam
    -c other ~/input.txt        
    -n pair1 'Pair 1 inputs'        
    -d pair1 'Pair 1 description'   
    -t pair1 directory             
    -e factor 5
    -m my-bowtie2
    -n 'Bowtie2 test'
    -v /tmp /opt
    -l 'foo: bar'
    -d 'Testings an example of using `funnel run` for a bowtie2 command'
  TPL)

  align_tpl=$(cat <<TPL
    'bowtie2 -f $factor -x $other -p1 $pair1 -p2 $pair2 -o $alignments' 
    -c other ~/input.txt        
    -n pair1 'Pair 1 inputs'        
    -d pair1 'Pair 1 description'   
    -t pair1 directory             
    -e factor 5
    -m my-bowtie2
    -n 'Bowtie2 test'
    -v /tmp /opt
    -l 'foo: bar'
    -d 'Testings an example of using `funnel run` for a bowtie2 command'
  TPL)

  align_sample_1=$(cat <<TPL
    -i pair1 ~/sample1/pair1.fastq            
    -i pair2 ~/sample1/pair2.fastq           
    -o alignments gs://bkt/bowtie/sample1/alignments.bam
  TPL)
  align_sample_2=$(cat <<TPL
    -i pair1 ~/sample2/pair1.fastq            
    -i pair2 ~/sample2/pair2.fastq           
    -o alignments gs://bkt/bowtie/sample2/alignments.bam
  TPL)

  sample_1_task=$(funnel run $align_tpl $align_sample_1)
  sample_2_task=$(funnel run $align_tpl $align_sample_2)

  funnel wait $sample_1_task $sample_2_task

  sample_align_tasks=$( for sample in $(ls -1 samples/); \
    funnel run $align_tpl \
      -i pair1 "samples/$sample/pair1.fastq" \
      -i pair2 "samples/$sample/pair2.fastq" \
      -o alignments "gs://bkt/bowtie/$sample/alignments.bam"; \
  done)

  funnel wait $sample_align_tasks
```

**Comment by jbingham** [comment]
Great. dsub has similar resource flags for cpu, ram, zones, etc.

For environment variables, we use the --env flag: eg, --env factor=5.
For folders, we use --input-recursive pair1=gs://b/path or --output-recursive outputFolder=gs://b/path

We can make it more concise by offering -i and -o for files, -I and -O for folders, -e for env vars, etc.

We don't currently support multiple disks -- we just create a single default disk, since our internal users have never wanted more than one so far. Maybe eventually we'll have users who want to mount an existing disk, or have more than one. Just not yet.

**Comment by buchanae** [comment]
Nice, ya, you highlighted most of the issues in that draft. I hope we can find a more automatic way of handling directories, but your suggestions are a great start.

Funnel isn't very smart about multiple disks either, since it's kinda hard to align easy configuration for that, so a single disk is fine for now.

**Comment by buchanae** [comment]
What do you think about hacking together simple workflows in this way? I'm hoping people will find it a useful alternative to a full-blown workflow engine. With some simple task caching, a funnel/dsub workflow could be run multiple times during development without spurious use of comments :)

**Comment by jbingham** [comment]
Yes. That's exactly how our internal users do it. For simple linear cases, or even some sharded stuff, it works just fine.

Sometimes we create a bash script like my_linear_workflow.sh:

#/bin/bash
dsub ...
dsub...

Then we run:

dsub --script my_linear_workflow.sh

That way we can close our laptops :)

**Comment by buchanae** [comment]
#98 added `funnel run` which covers a lot of what was discussed here, but there are some left overs in #105 and probably additional refinement over time.

**Comment by buchanae** [comment]
In summary, here's an idea of what funnel run supports

```
Usage:
  funnel run 'CMD' [flags]

General flags:
  -S, --server      Address of Funnel server. Default: http://localhost:8000
  -c, --container   Containter the command runs in. Default: alpine
  -p, --print       Print the task without running it.
      --scatter     Scatter multiple tasks, one per row of the given file.
      --wait        Wait for the task to finish before exiting.
      --wait-for    Wait for the given task IDs before running the task.
  -x, --extra       Parse funnel run arguments from the given string.
  -X, --extra-file  Parse funnel run arguments from the given file.

Input/output file flags:
  -i, --in          Input file e.g. varname=/path/to/input.txt
  -I, --in-dir      Input directory e.g. varname=/path/to/dir
  -o, --out         Output file e.g. varname=/path/to/output.txt
  -O, --out-dir     Output directory e.g. varname=/path/to/dir
  -C, --contents    Include input file contents a file e.g. varname=/path/to/in.txt
      --stdin       File to write to stdin to the command.
      --stdout      File to write to stdout of the command.
      --stderr      File to write to stderr of the command.

Resource request flags:
      --cpu         Number of CPUs to request.
      --ram         Amount of RAM to request, in GB.
      --disk        Amount of disk space to request, in GB.
      --zone        Require task be scheduled in certain zones.
      --preemptible Allow task to be scheduled on preemptible workers.

Other flags:
  -n, --name         Task name.
      --description  Task description.
      --project      Associated project ID.
      --tag          Arbitrary key-value tags, e.g. tagname=tagvalue
  -e, --env          Environment variables, e.g. envvar=foo
  -w, --workdir      Containter working directory.
      --vol          Define a volume on the container.

Examples:
  # Simple md5sum of a file and save the stdout.
  funnel run 'md5sum $in' -i in=input.txt --stdout output.txt

  # Use a different container.
  funnel run 'echo hello world' -c ubuntu

  # Print the task JSON instead of running it.
  funnel run 'echo $in" -i in=input.txt --print

  # md5sum all files in a directory.
  funnel run 'md5sum $d' -I d=./inputs --stdout output.txt

  # Sleep for 5 seconds, and wait for the sleep to finish.
  funnel run 'sleep 5' --wait

  # Reuse a set of arguments across multiple runs.
  args='--wait --cpu 10 --ram 60 --disk 2000'
  funnel run 'echo hello world" -x $args
  funnel run 'echo hello world again" -x $args

  # Set environment variables
  funnel run 'echo $MSG' -e MSG=Hello

  # When writing lots of arguments, Bash heredoc can be helpful.
  funnel run 'myprog -a $argA -b $argB -i $file1 -d $dir1' <<ARGS
    --container myorg/mycontainer
    --name 'MyProg test'
    --description 'Funnel run example of writing many arguments.'
    --tag meta=val
    --in file1=input.txt
    --in-dir dir1=inputs
    --stdout /path/to/stdout.txt
    --stderr /path/to/stderr.txt
    --env argA=1
    --env argB=2
    --vol /tmp
    --vol /opt
    --cpu 8
    --ram 32
    --disk 100
    --preemptible
    --zone us-west1-a
  ARGS
```
